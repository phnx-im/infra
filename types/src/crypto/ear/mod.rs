// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! This module and its submodules contain structs and types to facilitate (EAR)
//! encryption of other structs on the backend. See the individual submodules
//! for details.

pub mod keys;
mod traits;

use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};
pub use traits::{
    EarDecryptable, EarEncryptable, EarKey, GenericDeserializable, GenericSerializable,
};

use aes_gcm::Aes256Gcm;
use serde::{Deserialize, Serialize};

/// This type determines the AEAD scheme used for encryption at rest (EAR) by
/// the backend.
/// TODO: Replace with a key-committing scheme.
pub type Aead = Aes256Gcm;
/// Key size of the [`Aead`] scheme
pub const AEAD_KEY_SIZE: usize = 32;
const AEAD_NONCE_SIZE: usize = 12;

// Convenience struct that allows us to keep ciphertext and nonce together.
#[derive(
    Clone,
    Debug,
    PartialEq,
    Eq,
    Serialize,
    Deserialize,
    TlsSerialize,
    TlsDeserializeBytes,
    TlsSize,
    sqlx::Type,
)]
#[sqlx(type_name = "aead_ciphertext")]
pub struct Ciphertext {
    ciphertext: Vec<u8>,
    nonce: [u8; AEAD_NONCE_SIZE],
}

impl Default for Ciphertext {
    fn default() -> Self {
        Self {
            ciphertext: vec![],
            nonce: [0u8; AEAD_NONCE_SIZE],
        }
    }
}

impl Ciphertext {
    pub fn new(ciphertext: Vec<u8>, nonce: [u8; AEAD_NONCE_SIZE]) -> Self {
        Self { ciphertext, nonce }
    }

    pub fn into_parts(self) -> (Vec<u8>, [u8; AEAD_NONCE_SIZE]) {
        (self.ciphertext, self.nonce)
    }
}

#[cfg(any(feature = "test_utils", test))]
impl Ciphertext {
    pub fn dummy() -> Self {
        Self {
            ciphertext: vec![1u8; 32],
            nonce: [1u8; AEAD_NONCE_SIZE],
        }
    }

    pub fn random() -> Self {
        use rand::Rng;

        let mut rng = rand::thread_rng();
        Self {
            ciphertext: rng.r#gen::<[u8; 32]>().into(),
            nonce: rng.r#gen::<[u8; AEAD_NONCE_SIZE]>(),
        }
    }

    pub fn flip_bit(&mut self) {
        let byte = self.ciphertext.pop().unwrap();
        self.ciphertext.push(byte ^ 1);
    }
}
