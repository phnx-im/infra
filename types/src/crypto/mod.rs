// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! A module that provides various traits and structs that allow the use of
//! cryptographic primitives such as AEAD, MACs and KDFs.
//!
//! TODO: Once const-generics allows the use of enums, we could get rid of a
//! number of structs and boilerplate code.
//! TODO: A proper RNG provider for use with all crypto functions that require
//! randomness, i.e. mainly secret and nonce sampling.
#![allow(unused_variables)]
use std::marker::PhantomData;

use hpke::{DecryptionKey, EncryptionPublicKey};
use serde::{Deserialize, Serialize};
use sha2::Sha256;
use thiserror::Error;
use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};

use crate::{LibraryError, crypto::ear::EarEncryptable, messages::QueueMessage};

use self::{
    ear::{Ciphertext, EarDecryptable, keys::RatchetKey},
    errors::RandomnessError,
    hpke::{HpkeDecryptionKey, HpkeEncryptionKey},
    kdf::{KdfDerivable, keys::RatchetSecret},
};

pub use opaque::OpaqueCiphersuite;

/// This type determines the hash function used by the backend.
pub type Hash = Sha256;

pub mod ear;
pub mod errors;
pub mod hpke;
pub mod kdf;
pub mod mac;
pub mod opaque;
pub mod ratchet;
pub mod secrets;
pub(super) mod serde_arrays;
pub mod signatures;

pub type RatchetKeyUpdate = Vec<u8>;

#[derive(
    Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TlsSerialize, TlsDeserializeBytes, TlsSize,
)]
pub struct RatchetEncryptionKey(EncryptionPublicKey);

impl RatchetEncryptionKey {
    #[cfg(any(test, feature = "test_utils"))]
    pub fn new_for_test(encryption_key: EncryptionPublicKey) -> Self {
        Self(encryption_key)
    }
}

#[derive(Clone, Serialize, Deserialize, sqlx::Type)]
#[sqlx(transparent)]
pub struct RatchetDecryptionKey(DecryptionKey);

impl RatchetDecryptionKey {
    pub fn generate() -> Result<Self, RandomnessError> {
        Ok(Self(DecryptionKey::generate()?))
    }

    pub fn encryption_key(&self) -> RatchetEncryptionKey {
        RatchetEncryptionKey(self.0.public_key().clone())
    }
}

#[derive(
    Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TlsSerialize, TlsDeserializeBytes, TlsSize,
)]
pub struct ConnectionEncryptionKey {
    encryption_key: EncryptionPublicKey,
}

impl AsRef<EncryptionPublicKey> for ConnectionEncryptionKey {
    fn as_ref(&self) -> &EncryptionPublicKey {
        &self.encryption_key
    }
}

impl HpkeEncryptionKey for ConnectionEncryptionKey {}

#[derive(Clone, Serialize, Deserialize)]
pub struct ConnectionDecryptionKey {
    decryption_key: DecryptionKey,
}

impl AsRef<DecryptionKey> for ConnectionDecryptionKey {
    fn as_ref(&self) -> &DecryptionKey {
        &self.decryption_key
    }
}

impl HpkeDecryptionKey for ConnectionDecryptionKey {}

impl ConnectionDecryptionKey {
    pub fn generate() -> Result<Self, RandomnessError> {
        Ok(Self {
            decryption_key: DecryptionKey::generate()?,
        })
    }

    pub fn encryption_key(&self) -> ConnectionEncryptionKey {
        ConnectionEncryptionKey {
            encryption_key: self.decryption_key.public_key().clone(),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::codec::PhnxCodec;

    use super::*;

    #[test]
    fn encryption_key_serde_codec() {
        let key = RatchetEncryptionKey::new_for_test(EncryptionPublicKey::from(vec![1, 2, 3]));
        insta::assert_binary_snapshot!(".cbor", PhnxCodec::to_vec(&key).unwrap());
    }

    #[test]
    fn encryption_key_serde_json() {
        let key = RatchetEncryptionKey::new_for_test(EncryptionPublicKey::from(vec![1, 2, 3]));
        insta::assert_json_snapshot!(key);
    }
}
