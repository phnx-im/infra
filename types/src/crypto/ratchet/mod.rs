// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use errors::{DecryptionError, EncryptionError};
use serde::de::DeserializeOwned;

use super::{errors::RandomnessError, *};

#[cfg(test)]
mod tests;

pub trait RatchetPayload<Ciphertext: RatchetCiphertext>:
    EarEncryptable<RatchetKey, Ciphertext> + EarDecryptable<RatchetKey, Ciphertext>
{
}
pub trait RatchetCiphertext: AsRef<Ciphertext> + From<Ciphertext> {}

impl<Ciphertext: RatchetCiphertext, T> RatchetPayload<Ciphertext> for T where
    T: EarEncryptable<RatchetKey, Ciphertext> + EarDecryptable<RatchetKey, Ciphertext>
{
}
impl<T> RatchetCiphertext for T where
    T: AsRef<Ciphertext> + From<Ciphertext> + Serialize + DeserializeOwned
{
}

// WARNING: If this struct is changed its implementation of ToSql and FromSql in the sqlite module
// must be updated and a new `QueueRatchetVersion` introduced.
#[derive(Serialize, Deserialize, Clone, Debug, TlsSerialize, TlsDeserializeBytes, TlsSize)]
pub struct QueueRatchet<Ciphertext: RatchetCiphertext, Payload: RatchetPayload<Ciphertext>> {
    sequence_number: u64,
    secret: RatchetSecret,
    key: RatchetKey,
    _phantom: PhantomData<(Ciphertext, Payload)>,
}

#[cfg(feature = "test_utils")]
impl<Ciphertext: RatchetCiphertext, Payload: RatchetPayload<Ciphertext>> PartialEq
    for QueueRatchet<Ciphertext, Payload>
{
    fn eq(&self, other: &Self) -> bool {
        self.sequence_number == other.sequence_number
            && self.secret == other.secret
            && self.key == other.key
            && self._phantom == other._phantom
    }
}

#[cfg(feature = "test_utils")]
impl<Ciphertext: RatchetCiphertext, Payload: RatchetPayload<Ciphertext>> Eq
    for QueueRatchet<Ciphertext, Payload>
{
}

impl<Ciphertext: RatchetCiphertext, Payload: RatchetPayload<Ciphertext>> TryFrom<RatchetSecret>
    for QueueRatchet<Ciphertext, Payload>
{
    type Error = LibraryError;

    fn try_from(secret: RatchetSecret) -> Result<Self, Self::Error> {
        let key = RatchetKey::derive(&secret, Vec::new()).map_err(|_| LibraryError)?;
        Ok(Self {
            sequence_number: 0,
            secret,
            key,
            _phantom: PhantomData,
        })
    }
}

// TODO: Implement the ratchet key.
impl<Ciphertext: RatchetCiphertext, Payload: RatchetPayload<Ciphertext>>
    QueueRatchet<Ciphertext, Payload>
{
    /// Initialize a new ratchet key.
    pub fn random() -> Result<Self, RandomnessError> {
        let secret = RatchetSecret::random()?;
        secret
            .try_into()
            .map_err(|_| RandomnessError::InsufficientRandomness)
    }

    fn ratchet_forward(&mut self) -> Result<(), EncryptionError> {
        let secret = RatchetSecret::derive(&self.secret, Vec::new())
            .map_err(|_| EncryptionError::SerializationError)?;
        let key = RatchetKey::derive(&secret, Vec::new())
            .map_err(|_| EncryptionError::SerializationError)?;

        self.secret = secret;
        self.key = key;
        self.sequence_number += 1;

        Ok(())
    }

    /// Encrypt the given payload.
    pub fn encrypt(&mut self, payload: Payload) -> Result<QueueMessage, EncryptionError> {
        // TODO: We want domain separation: FQDN, UserID & ClientID.
        let ciphertext = payload.encrypt(&self.key)?;

        let queue_message = QueueMessage {
            sequence_number: self.sequence_number,
            ciphertext: ciphertext.as_ref().clone(),
        };

        self.ratchet_forward()?;

        Ok(queue_message)
    }

    /// Decrypt the given payload.
    pub fn decrypt(&mut self, queue_message: QueueMessage) -> Result<Payload, DecryptionError> {
        let ciphertext = queue_message.ciphertext.into();
        let plaintext = Payload::decrypt(&self.key, &ciphertext)?;
        self.ratchet_forward()
            .map_err(|_| DecryptionError::DecryptionError)?;
        Ok(plaintext)
    }

    /// Sample some fresh entropy and inject it into the current key. Returns the entropy.
    pub fn update(&mut self) -> RatchetKeyUpdate {
        todo!()
    }

    /// Get the current sequence number
    pub fn sequence_number(&self) -> u64 {
        self.sequence_number
    }

    pub fn secret(&self) -> &RatchetSecret {
        &self.secret
    }

    pub fn key(&self) -> &RatchetKey {
        &self.key
    }
}

mod sqlite {
    use sqlx::{Database, Decode, Encode, Sqlite, Type, encode::IsNull, error::BoxDynError};

    use crate::codec::PhnxCodec;

    use super::*;

    // When adding a variant to this enum, the new variant must be called
    // `CurrentVersion` and the current version must be renamed to `VX`, where `X`
    // is the next version number. The content type of the old `CurrentVersion` must
    // be renamed and otherwise preserved to ensure backwards compatibility.
    #[derive(Serialize, Deserialize)]
    enum VersionedQueueRatchet {
        CurrentVersion(Vec<u8>),
    }

    impl<Ciphertext: RatchetCiphertext, Payload: RatchetPayload<Ciphertext>> Type<Sqlite>
        for QueueRatchet<Ciphertext, Payload>
    {
        fn type_info() -> <Sqlite as Database>::TypeInfo {
            <Vec<u8> as Type<Sqlite>>::type_info()
        }
    }

    impl<Ciphertext: RatchetCiphertext, Payload: RatchetPayload<Ciphertext>> Encode<'_, Sqlite>
        for QueueRatchet<Ciphertext, Payload>
    {
        fn encode_by_ref(
            &self,
            buf: &mut <Sqlite as Database>::ArgumentBuffer<'_>,
        ) -> Result<IsNull, BoxDynError> {
            let ratchet_bytes = PhnxCodec::to_vec(self)?;
            let versioned_ratchet_bytes =
                PhnxCodec::to_vec(&VersionedQueueRatchet::CurrentVersion(ratchet_bytes))?;
            Encode::<Sqlite>::encode(versioned_ratchet_bytes, buf)
        }
    }

    impl<Ciphertext: RatchetCiphertext, Payload: RatchetPayload<Ciphertext>> Decode<'_, Sqlite>
        for QueueRatchet<Ciphertext, Payload>
    {
        fn decode(value: <Sqlite as Database>::ValueRef<'_>) -> Result<Self, BoxDynError> {
            let bytes: &[u8] = Decode::<Sqlite>::decode(value)?;
            let VersionedQueueRatchet::CurrentVersion(ratchet_bytes) =
                PhnxCodec::from_slice(bytes)?;
            let ratchet = PhnxCodec::from_slice(&ratchet_bytes)?;
            Ok(ratchet)
        }
    }
}

#[cfg(test)]
mod test {
    use crate::{
        codec::PhnxCodec,
        messages::{EncryptedQsQueueMessage, client_ds::QsQueueMessagePayload},
    };

    use super::*;

    // Note: the type parameters are not important
    fn queue_ratchet() -> QueueRatchet<EncryptedQsQueueMessage, QsQueueMessagePayload> {
        let secret: &[u8; 32] = b"abcdefghijklmnopqrstuvwxyz012345";
        let ratchet_secret = RatchetSecret::new_for_test((*secret).into());
        QueueRatchet::try_from(ratchet_secret).unwrap()
    }

    #[test]
    fn test_queue_ratchet_serde_codec() {
        insta::assert_binary_snapshot!(".cbor", PhnxCodec::to_vec(&queue_ratchet()).unwrap());
    }

    #[test]
    fn test_queue_ratchet_serde_json() {
        insta::assert_json_snapshot!(queue_ratchet());
    }
}
