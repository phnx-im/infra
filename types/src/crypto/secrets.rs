// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! This module contains the [`Secret`] struct, which is meant to be
//! accessible only within the [`crate::crypto`] module. This
//! module should stay private, such that the [`Secret`] struct can stay public
//! and the type checker happy.
use std::{fmt::Display, ops::Deref};

use rand::{RngCore, SeedableRng};
use secrecy::{
    CloneableSecret, SerializableSecret,
    zeroize::{Zeroize, ZeroizeOnDrop},
};
use serde::{Deserialize, Serialize};
use sqlx::{Database, Decode, Encode, Sqlite, Type, encode::IsNull, error::BoxDynError};
use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};

use super::RandomnessError;

/// Struct that contains a (symmetric) secret of fixed length LENGTH.
#[derive(
    TlsSerialize, TlsDeserializeBytes, TlsSize, Clone, PartialEq, Eq, Serialize, Deserialize,
)]
pub struct Secret<const LENGTH: usize> {
    #[serde(with = "super::serde_arrays")]
    secret: [u8; LENGTH],
}

impl<const LENGTH: usize> From<[u8; LENGTH]> for Secret<LENGTH> {
    fn from(secret: [u8; LENGTH]) -> Self {
        Self { secret }
    }
}

impl<const LENGTH: usize> Secret<LENGTH> {
    /// Get the internal secret value
    pub fn secret(&self) -> &[u8; LENGTH] {
        &self.secret
    }

    pub(super) fn into_secret(self) -> [u8; LENGTH] {
        self.secret
    }

    /// Generate a fresh, random secret.
    pub(super) fn random() -> Result<Self, RandomnessError> {
        let mut secret = [0; LENGTH];
        // TODO: Use a proper rng provider.
        rand_chacha::ChaCha20Rng::from_entropy()
            .try_fill_bytes(secret.as_mut_slice())
            .map_err(|_| RandomnessError::InsufficientRandomness)?;
        Ok(Self { secret })
    }
}

// Ensure that secrets are wiped from memory securely upon being dropped.
impl<const LENGTH: usize> Zeroize for Secret<LENGTH> {
    fn zeroize(&mut self) {
        self.secret.zeroize()
    }
}

impl<const LENGTH: usize> ZeroizeOnDrop for Secret<LENGTH> {}

// Ensures that secrets are not printed in debug outputs.
impl<const LENGTH: usize> std::fmt::Debug for Secret<LENGTH> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Secret: [[REDACTED]]").finish()
    }
}

// Ensures that secrets are not printed in format strings.
impl<const LENGTH: usize> Display for Secret<LENGTH> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[[REDACTED]]")
    }
}

impl<const LENGTH: usize> Type<Sqlite> for Secret<LENGTH> {
    fn type_info() -> <Sqlite as Database>::TypeInfo {
        <&[u8] as Type<Sqlite>>::type_info()
    }
}

impl<'q, const LENGTH: usize> Encode<'q, Sqlite> for Secret<LENGTH> {
    fn encode_by_ref(
        &self,
        buf: &mut <Sqlite as Database>::ArgumentBuffer<'q>,
    ) -> Result<IsNull, BoxDynError> {
        let bytes: Box<[u8]> = self.secret.into();
        Encode::<Sqlite>::encode(bytes, buf)
    }
}

impl<'r, const LENGTH: usize> Decode<'r, Sqlite> for Secret<LENGTH> {
    fn decode(value: <Sqlite as Database>::ValueRef<'r>) -> Result<Self, BoxDynError> {
        let bytes: &[u8] = Decode::<Sqlite>::decode(value)?;
        Ok(Secret {
            secret: bytes.try_into()?,
        })
    }
}

#[derive(Clone, Serialize, Deserialize, sqlx::Type)]
#[sqlx(transparent)]
pub(super) struct SecretBytes(Vec<u8>);

impl From<Vec<u8>> for SecretBytes {
    fn from(secret: Vec<u8>) -> Self {
        Self(secret)
    }
}

impl Deref for SecretBytes {
    type Target = Vec<u8>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Zeroize for SecretBytes {
    fn zeroize(&mut self) {
        self.0.zeroize();
    }
}

impl ZeroizeOnDrop for SecretBytes {}

// Ensures that secrets are not printed in debug outputs.
impl std::fmt::Debug for SecretBytes {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Secret: [[REDACTED]]").finish()
    }
}

// Ensures that secrets are not printed in format strings.
impl Display for SecretBytes {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[[REDACTED]]")
    }
}

impl SerializableSecret for SecretBytes {}
impl CloneableSecret for SecretBytes {}
