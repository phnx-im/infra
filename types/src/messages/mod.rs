// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::fmt;

use mls_assist::{
    openmls::prelude::{KeyPackage, KeyPackageIn, OpenMlsRand},
    openmls_rust_crypto::OpenMlsRustCrypto,
    openmls_traits::OpenMlsProvider,
};
use serde::{Deserialize, Serialize};
use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize, TlsVarInt};

use crate::crypto::{
    ear::{Ciphertext, EarDecryptable, EarEncryptable, keys::KeyPackageEarKey},
    errors::RandomnessError,
};

pub mod client_as;
pub mod client_as_out;
pub mod client_ds;
pub mod client_ds_out;
pub mod client_qs;
pub mod client_qs_out;
pub mod push_token;
pub mod welcome_attribution_info;

#[derive(
    Serialize,
    Deserialize,
    TlsSerialize,
    TlsDeserializeBytes,
    TlsSize,
    PartialEq,
    Eq,
    Clone,
    Debug,
    sqlx::Type,
)]
#[sqlx(transparent)]
pub struct FriendshipToken(Vec<u8>);

impl FriendshipToken {
    pub fn random() -> Result<Self, RandomnessError> {
        let token = OpenMlsRustCrypto::default()
            .rand()
            .random_vec(32)
            .map_err(|_| RandomnessError::InsufficientRandomness)?;

        Ok(Self(token))
    }

    #[cfg(test)]
    pub(crate) fn new_for_test(token: Vec<u8>) -> Self {
        Self(token)
    }

    pub fn token(&self) -> &[u8] {
        self.0.as_ref()
    }
}

/// Enum encoding the version of the MlsInfra protocol that was used to create
/// the given message.
#[derive(
    Debug,
    TlsSerialize,
    TlsDeserializeBytes,
    TlsSize,
    Clone,
    Copy,
    PartialEq,
    Eq,
    Serialize,
    Deserialize,
)]
#[repr(u8)]
pub enum MlsInfraVersion {
    Alpha,
}

impl Default for MlsInfraVersion {
    fn default() -> Self {
        Self::Alpha
    }
}

// === Queue ===

#[derive(
    Clone, Debug, PartialEq, Eq, Serialize, Deserialize, TlsSerialize, TlsDeserializeBytes, TlsSize,
)]
pub struct QueueMessage {
    pub sequence_number: u64,
    pub ciphertext: Ciphertext,
}

#[derive(
    Clone, Debug, PartialEq, Serialize, Deserialize, TlsSerialize, TlsDeserializeBytes, TlsSize,
)]
pub struct EncryptedQsQueueMessage {
    payload: Ciphertext,
}

impl From<Ciphertext> for EncryptedQsQueueMessage {
    fn from(payload: Ciphertext) -> Self {
        Self { payload }
    }
}

impl AsRef<Ciphertext> for EncryptedQsQueueMessage {
    fn as_ref(&self) -> &Ciphertext {
        &self.payload
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, TlsSerialize, TlsDeserializeBytes, TlsSize)]
pub struct EncryptedAsQueueMessage {
    payload: Ciphertext,
}

impl From<Ciphertext> for EncryptedAsQueueMessage {
    fn from(payload: Ciphertext) -> Self {
        Self { payload }
    }
}

impl AsRef<Ciphertext> for EncryptedAsQueueMessage {
    fn as_ref(&self) -> &Ciphertext {
        &self.payload
    }
}

#[derive(Debug, TlsDeserializeBytes, TlsSerialize, TlsSize)]
#[repr(u8)]
pub enum AsTokenType {
    AsEnqueue,
    DsGroupCreation,
    DsGroupOperation,
}

/// Ciphertext that contains a KeyPackage and an intermediary client certficate.
/// TODO: do we want a key committing scheme here?
#[derive(
    Debug, TlsSerialize, TlsDeserializeBytes, TlsSize, Clone, Serialize, Deserialize, sqlx::Type,
)]
#[sqlx(transparent)]
pub struct QsEncryptedKeyPackage(Ciphertext);

impl AsRef<Ciphertext> for QsEncryptedKeyPackage {
    fn as_ref(&self) -> &Ciphertext {
        &self.0
    }
}

impl From<Ciphertext> for QsEncryptedKeyPackage {
    fn from(ctxt: Ciphertext) -> Self {
        Self(ctxt)
    }
}

impl EarDecryptable<KeyPackageEarKey, QsEncryptedKeyPackage> for KeyPackageIn {}
impl EarEncryptable<KeyPackageEarKey, QsEncryptedKeyPackage> for KeyPackage {}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ApiVersion(TlsVarInt);

impl ApiVersion {
    pub const fn new(version: u64) -> Option<Self> {
        // Note: At the moment of writing, Option::map is not a const fn.
        match TlsVarInt::new(version) {
            Some(v) => Some(Self(v)),
            None => None,
        }
    }

    pub const fn value(&self) -> u64 {
        self.0.value()
    }

    pub(crate) const fn from_tls_value(value: TlsVarInt) -> Self {
        Self(value)
    }

    pub const fn tls_value(&self) -> TlsVarInt {
        self.0
    }
}

impl fmt::Display for ApiVersion {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value())
    }
}

#[cfg(test)]
mod test {
    use crate::codec::PhnxCodec;

    use super::*;

    #[test]
    fn test_queue_message_serde_codec() {
        let message = QueueMessage {
            sequence_number: 1,
            ciphertext: Ciphertext::dummy(),
        };
        insta::assert_binary_snapshot!(".cbor", PhnxCodec::to_vec(&message).unwrap());
    }

    #[test]
    fn test_queue_message_serde_json() {
        let message = QueueMessage {
            sequence_number: 1,
            ciphertext: Ciphertext::dummy(),
        };
        insta::assert_json_snapshot!(message);
    }
}
