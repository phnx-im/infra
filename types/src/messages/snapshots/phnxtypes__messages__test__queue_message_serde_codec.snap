---
source: types/src/messages/mod.rs
expression: "PhnxCodec::to_vec(&message).unwrap()"
extension: cbor
snapshot_kind: binary
---
